flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncWaitOperator.java

```diff
@@ -0,0 +1,224 @@
+		this.inStreamElementSerializerForTest = serializer;
+	}
+
+	@Override
+	public void setup(StreamTask<?, ?> containingTask, StreamConfig config, Output<StreamRecord<OUT>> output) {
+		super.setup(containingTask, config, output);
+
+		if (this.inStreamElementSerializerForTest != null) {
+			this.inStreamElementSerializer = this.inStreamElementSerializerForTest;
+		}
+		else {
+			this.inStreamElementSerializer = new StreamElementSerializer(this.getOperatorConfig().getTypeSerializerIn1(getUserCodeClassloader()));
+		}
+
+		this.collector = new TimestampedCollector<>(output);
+
+		this.checkpointLock = containingTask.getCheckpointLock();
+
+		this.buffer = new AsyncCollectorBuffer<>(bufferSize, mode, output, collector, this.checkpointLock, this);
+
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+
+		// process stream elements from state
+		if (this.recoveredStreamElements != null) {
+			for (StreamElement element : this.recoveredStreamElements.get()) {
+				if (element.isRecord()) {
+					processElement(element.<IN>asRecord());
+				}
+				else if (element.isWatermark()) {
+					processWatermark(element.asWatermark());
+				}
+				else if (element.isLatencyMarker()) {
+					processLatencyMarker(element.asLatencyMarker());
+				}
+				else {
+					throw new Exception("Unknown record type: "+element.getClass());
+				}
+			}
+			this.recoveredStreamElements = null;
+		}
+
+
+		if (emitFlag) {
+			buffer.startEmitterThread();
+		}
+	}
+
+	@Override
+	public void processElement(StreamRecord<IN> element) throws Exception {
+		AsyncCollector<IN, OUT> collector = buffer.addStreamRecord(element);
+
+		userFunction.asyncInvoke(element.getValue(), collector);
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		buffer.addWatermark(mark);
+	}
+
+	@Override
+	public void processLatencyMarker(LatencyMarker latencyMarker) throws Exception {
+		buffer.addLatencyMarker(latencyMarker);
+	}
+
+	@Override
+	public void snapshotState(StateSnapshotContext context) throws Exception {
+		super.snapshotState(context);
+
+		ListState<StreamElement> partitionableState =
+				getOperatorStateBackend().getOperatorState(new ListStateDescriptor<>(STATE_NAME, inStreamElementSerializer));
+		partitionableState.clear();
+
+		Collection<StreamElement> elements = buffer.getStreamElementsInBuffer();
+		for (StreamElement element : elements) {
+			partitionableState.add(element);
+		}
+	}
+
+	@Override
+	public void initializeState(StateInitializationContext context) throws Exception {
+		recoveredStreamElements =
+				context.getManagedOperatorStateStore().getOperatorState(new ListStateDescriptor<>(STATE_NAME, inStreamElementSerializer));
+
+	}
+
+	@Override
+	public void close() throws Exception {
+		// for test only, we do not have to wait until buffer is empty.
+		if (!emitFlag) {
+			return;
+		}
+
+		buffer.waitEmpty();
+		buffer.stopEmitterThread();
```

Comment:
Emm, I think you are right. We'd better stop the emitter thread while closing the operator, so that the order to dispose operators can still be from tail to head - no modification to the code in `StreamTask.tryDisposeAllOperators()`.